### PR TITLE
Use the internal CUB memory pool for everything by default

### PR DESCRIPTION
**Pull request by benson31**
Unless you specify the mode, it will use "new" on CPUs and CUB on GPUs by default.

### CODE REVIEW

**Comment by benson31** [comment]
Crap don't review or merge this. Something bad happened to my git history I think...

**Comment by benson31** [comment]
Actually it looks ok.

@bvanessen Can you add @timmoon10 to whatever list he has to be on (LLNL org, LBANN team, llnl/elemental admins, whatever) so he can do things in the repo and, most importantly, actually show up in the user-mention auto-complete list and in the suggested reviewers list... Thanks!

**Comment by benson31** [comment]
@bvanessen Same request for @ndryden since I can't find him either!

File: include/El/core/SimpleBuffer.hpp [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+};// struct MemorySetter<Device::CPU>
+
+#ifdef HYDROGEN_HAVE_CUDA
+template <>
+struct MemorySetter<Device::GPU>
+{
+    template<typename T>
+    static void setBufferToValue(T* buffer, size_t size, T const& value)
+    {
+        if( value == T(0) )
```
Comment (timmoon10) [inline include/El/core/SimpleBuffer.hpp:67]: We could probably use the `Fill_GPU_impl()` in src/blas_like/level1/GPU/Fill.cu to do this entirely on GPU.

**Review by timmoon10 — APPROVED** [review]
Looks reasonable to me. I'd just like to make explicit the trade-offs we're making:
- Using the CUB memory pool gives us cheap memory allocation/deallocation. We don't need to think very hard about temporary GPU matrices.
- We may see an effective reduction in memory capacity. The memory pool grows with a fixed growth factor (the smallest is 2), so it will miss some memory if the available memory is not a power of two. In addition, repeated allocation/deallocation may make us susceptible to memory fragmentation, reducing our ability to make large memory allocations.

**Comment by ndryden** [comment]
It would be nice if they supported smaller growth factors... There's a good argument that φ=1.61... (the Golden Ratio) is an optimal growth factor, precisely to help deal with fragmentation. Typically rounded to 1.5 because computers.

As long as we can keep track of the memory usage, I don't think we need to be too concerned about a little extra space being used during training. If it becomes egregious, we can revisit this. (Ditto if we ever decide to optimize for inference.)

**Comment by timmoon10** [comment]
Maybe we can ask the Nvidia people very nicely for golden growth (or even better give us the option to put all GPU memory under control of the memory pool).

~~The main worry I have with memory fragmentation is unavoidable with a memory pool approach. I imagine the CUDA driver can do some stuff behind the scenes with the address space to make non-contiguous memory effectively contiguous for the user, but we lose that with the memory pool.~~ I now see that the memory pool maintains memory blocks of various sizes and rounds up an allocation to the nearest block size. I better understand why we want golden growth.

**Pull request merged**